crates/diagnostic/src/lib.rs

```diff
@@ -505,6 +505,42 @@ impl Diagnostic {
             }
         }
     }
+
+    /// hir_tyのチェックエラー情報を元に診断情報を作成します。
+    pub fn from_hir_ty_check_error(
+        db: &base_db::SalsaDatabase,
+        _hir_file: hir::HirFile,
+        _file_db: &hir::HirFileDatabase,
+        source_map: &hir::HirFileSourceMap,
+        error: &hir_ty::TypeCheckError,
+    ) -> Self {
+        match error {
+            hir_ty::TypeCheckError::UnresolvedType { expr } => {
+                let text_range = expr.text_range(db, source_map);
+                Diagnostic {
+                    file: source_map.file(db),
+                    title: "Unresolved type".to_string(),
+                    head_offset: text_range.start().into(),
+                    messages: vec![Message {
+                        message: "cannot resolve type".to_string(),
+                        range: text_range,
+                    }],
+                }
+            }
+            hir_ty::TypeCheckError::ImmutableReassignment { expr } => {
+                let text_range = expr.text_range(db, source_map);
+                Diagnostic {
+                    file: source_map.file(db),
+                    title: "Immutable reassignment".to_string(),
+                    head_offset: text_range.start().into(),
+                    messages: vec![Message {
+                        message: "cannot reassign to immutable variable".to_string(),
+                        range: text_range,
+                    }],
+                }
+            }
+        }
+    }
```

Comment:
**Optimization Issue**: The new function 'from_hir_ty_check_error' introduces handling for 'TypeCheckError' variants such as 'UnresolvedType' and 'ImmutableReassignment'. However, the implementation repeats the pattern of fetching 'text_range' and constructing 'Diagnostic' objects in a way that could be abstracted to reduce redundancy and improve maintainability. <br> **Fix**: Refactor the repeated logic into a private helper function that constructs the 'Diagnostic' object. This function should take parameters necessary to customize the 'Diagnostic' for different error types. <br> **Code Suggestion**: 
 ```
 +    /// Helper to create diagnostic from type check errors
 +    fn create_type_check_diagnostic(
 +        db: &base_db::SalsaDatabase,
 +        source_map: &hir::HirFileSourceMap,
 +        expr: &hir::Expr,
 +        error_title: String,
 +        error_message: String,
 +    ) -> Diagnostic {
 +        let text_range = expr.text_range(db, source_map);
 +        Diagnostic {
 +            file: source_map.file(db),
 +            title: error_title,
 +            head_offset: text_range.start().into(),
 +            messages: vec![Message {
 +                message: error_message,
 +                range: text_range,
 +            }],
 +        }
 +
 +    pub fn from_hir_ty_check_error(
 +        db: &base_db::SalsaDatabase,
 +        _hir_file: hir::HirFile,
 +        _file_db: &hir::HirFileDatabase,
 +        source_map: &hir::HirFileSourceMap,
 +        error: &hir_ty::TypeCheckError,
 +    ) -> Self {
 +        match error {
 +            hir_ty::TypeCheckError::UnresolvedType { expr } => {
 +                Self::create_type_check_diagnostic(
 +                    db,
 +                    source_map,
 +                    expr,
 +                    "Unresolved type".to_string(),
 +                    "cannot resolve type".to_string(),
 +                )
 +            },
 +            hir_ty::TypeCheckError::ImmutableReassignment { expr } => {
 +                Self::create_type_check_diagnostic(
 +                    db,
 +                    source_map,
 +                    expr,
 +                    "Immutable reassignment".to_string(),
 +                    "cannot reassign to immutable variable".to_string(),
 +                )
 +            },
 +        }
 +    }
 ```